memcache/selector_test.go

```diff
@@ -16,24 +16,284 @@ limitations under the License.
+func functionalTest(t *testing.T, ports []string) {
+	srvs := []string{}
+	for _, port := range ports {
+		srvs = append(srvs, fmt.Sprintf("localhost:%s", port))
+	}
+	client := New(srvs...)
+
+	wg := sync.WaitGroup{}
+
+	server := func(port string) {
+		for i := 0; i < 5; i++ {
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second*3)
+			defer cancel()
+			cmd := exec.CommandContext(ctx, "memcached", "-p", port)
+			err := cmd.Run()
+			assertEqualError(t, err, "signal: killed")
+
+			// no randomisation - keep approximately in lockstep to keep the amount of downtime high
+			time.Sleep(time.Millisecond*500)
+		}
+		wg.Done()
+	}
+
+	done := make(chan struct{})
+	mx := sync.Mutex{}
+	resultCounts := map[string]int{}
+	networkErrorCategory := "[network error]"
+	internalTimeoutError := "[memcachepl timeout]"
+	writer := func() {
+		for {
+			select {
+			case <-done:
+				break
+			default:
+				err := client.Set(&Item{
+					Key:   "hi",
+					Value: []byte("val"),
+				})
+				errString := fmt.Sprintf("%v", err)
+				var netErr net.Error
+				if errors.As(err, &netErr) {
+					errString = networkErrorCategory
+				}
+				var cte *ConnectTimeoutError
+				if errors.As(err, &cte) {
+					errString = internalTimeoutError
+				}
+				mx.Lock()
+				resultCounts[errString]++
+				mx.Unlock()
+			}
+			time.Sleep(time.Duration(rand.Int31n(5)) * time.Millisecond)
+		}
+	}
+
+	// start servers
+	wg.Add(len(srvs))
+	for _, port := range ports {
+		go server(port)
+	}
+
+	go (func() {
+		wg.Wait()
+		close(done)
+	})()
+
+	// start writers
+	go writer()
+	go writer()
+	go writer()
+	go writer()
+	go writer()
+	go writer()
+
+	// wait for test run to end
+	wg.Wait()
+
+	// avoid race-detection errors when we range over results
+	mx.Lock()
+	defer mx.Unlock()
+
+	unexpectedErrors := map[string]struct{}{}
+	total := 0
+	for errVal, count := range resultCounts {
+		total += count
+		switch errVal {
+
+		// filter out success, and expected errors
+		case "<nil>":
+		case "EOF":
+		case internalTimeoutError:
+		case networkErrorCategory:
+		case ErrNoServers.Error():
+
+		default:
+			unexpectedErrors[errVal] = struct{}{}
 		}
 	}
+
+	t.Log("aggregate result of Set calls", resultCounts)
+
+	// we're up for approximately 2.5 seconds out of every three, so we should have a lot of successes too
+	assertGreaterOrEqual(t, resultCounts["<nil>"], total/10, "expected many successes")
+
+	for s := range unexpectedErrors {
+		t.Errorf("unexpected errors during test - either indicates a bug or something that needs categorisation: %s", s)
+	}
+
+	// test that we have greatly more breaker errors than network errors, reducing the reconnections during unavailability by a large factor
+	networkFails := resultCounts[networkErrorCategory] + resultCounts[internalTimeoutError] +
+		resultCounts["EOF"]
+	assertGreaterOrEqual(t, resultCounts[ErrNoServers.Error()], networkFails*10, "didn't reduce reconnection load as expected")
+}
```

Comment:
somehow I found the `10` number in both calls to `assertGreaterOrEqual` a little mysterious, and it wasn't clear to me that this is our heuristic for what the ratio of behaviours should be.

It may help to leave a comment to explain our reasoning a little more (I wasn't sure if we were having to do a little bit of math to convert between percentages and decimals, for example).

Reply:
Good point - I'll comment that this is a rough boundaries based on the observed behaviour with the amount of up and downtime specified in the test above.